Use new analyzer for EXPLAIN SYNTAX

vdimir opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* `EXPLAIN SYNTAX` now uses a new analyzer. It returns AST built from the query tree. Added option `query_tree_passes` to control the number of passes to executed before converting query tree to the AST.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

vdimir committed:
Use new analyzer for EXPLAIN SYNTAX

robot-clickhouse commented:
<!-- automatic status comment for PR #74536 from ClickHouse/ClickHouse:vimir/explain_syntax_use_new_analyze -->
*This is an automated comment for commit ce7faefd9ca51b16b69f8c19f7ea06f42ba4ba10 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/74536/ce7faefd9ca51b16b69f8c19f7ea06f42ba4ba10/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/74536/ce7faefd9ca51b16b69f8c19f7ea06f42ba4ba10/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

novikd commented:
It's the same as #63402. The problem is that it requires manually updating many tests because the results differ for the new and old analyzers.

vdimir commented:
> It's the same as #63402. The problem is that it requires manually updating many tests because the results differ for the new and old analyzers.

We can manually set `allow_experiemental_analyzer = 1` for `EXPLAIN SYNTAX` queries in tests and update referece, I'll try

novikd commented:
> We can manually set allow_experiemental_analyzer = 1 for EXPLAIN SYNTAX queries in tests and update referece

We can lose some AST-level optimizations without noticing. That's why I didn't proceed with my PR.

vdimir committed:
clickhouse-tests support reference files for old analyzer

vdimir committed:
add test 03312_explain_syntax_analyzer

vdimir committed:
canonize explain syntax result for new analyzer

vdimir committed:
update 02353_order_by_tuple

vdimir committed:
update 02420_final_setting_analyzer

vdimir committed:
update 02497_if_transform_strings_to_enum

vdimir committed:
update 03006_parallel_replicas_cte_explain_syntax_crash

vdimir committed:
update 03212_optimize_with_constraints_logical_error

vdimir committed:
update 03161_cnf_reduction

vdimir committed:
update 03212_optimize_with_constraints_logical_error

vdimir committed:
update 03274_udf_in_join

vdimir committed:
update 03282_memory_transaction_crash

vdimir committed:
update 01300_group_by_other_keys_having

vdimir commented:
@novikd 

I added support for using files with the `*.oldanalyzer.reference` extension instead of `.reference` when the old analyzer is enabled via `USE_OLD_ANALYZER` or the `client-option` in `clickhouse-test`.
Then I renamed the existing reference files for tests with `EXPLAIN SYNTAX` and generated new reference files for the new analyzer.

Around ten tests required additional fixes, which you can review in the `update ...` commits in this PR. These fixes mainly address cases where `enable_analyzer = 1` is explicitly set in the test, allowing us to disregard results for the old analyzer.

So, I'd suggest reviewing commit by commit ignoring huge `canonize explain syntax result for new analyzer` commit.

novikd commented on tests/queries/0_stateless/00597_push_down_predicate_long.reference:
```diff
@@ -68,67 +74,74 @@ FROM
-SELECT
-    id,
-    subquery
+SELECT *
```
That looks suspicious. `EXPLAIN SYNTAX` should produce query without matchers.

vdimir replied:
It's because we do not perform query tree analysys. In old analyzer we replace marchers on syntax level, in new one only if passes are run. I set running passes disabled by default because as you mention we want to keep `EXPLAIN SYNTAX` more user friendly (comparing to developer friendly `EXPLAIN QUERY TREE`), so it just performs syntax analysis without semantic analysis which is done in passes

novikd replied:
As I understand, the purpose of `EXPLAIN SYNTAX` is to see what the query looks like after semantic analysis and AST-level optimizations.

novikd commented on src/Interpreters/InterpreterFactory.cpp:
```diff
@@ -233,7 +233,7 @@ InterpreterFactory::InterpreterPtr InterpreterFactory::get(ASTPtr & query, Conte
     {
         const auto kind = query->as<ASTExplainQuery>()->getKind();
-        if (kind == ASTExplainQuery::ParsedAST || kind == ASTExplainQuery::AnalyzedSyntax)
+        if (kind == ASTExplainQuery::ParsedAST)
             context->setSetting("allow_experimental_analyzer", false);
```
Why do we need to disable the analyzer to show parsed AST?

vdimir replied:
I haven't tried to enable it (it's a different type of explain), probably it should be addressed separately

vdimir committed:
documentation EXPLAIN SYNTAX

novikd commented on src/Interpreters/InterpreterExplainQuery.cpp:
```diff
@@ -423,69 +489,45 @@ QueryPipeline InterpreterExplainQuery::executeImpl()
+            ContextPtr query_context = getContext();
+            if (query_context->getSettingsRef()[Setting::allow_experimental_analyzer])
+            {
+                bool explain_ok = explainQueryTree(ast.getExplainedQuery(), query_context, QueryTreeSettings{
+                    .run_passes = settings.run_query_tree_passes,
+                    .dump_tree = false,
+                    .dump_passes = false,
+                    .dump_ast = true,
+                    .passes = settings.query_tree_passes,
+                    .ast_one_line = settings.oneline,
+                }, buf);
```
It seems to be overcomplicated. `EXPLAIN QUERY TREE` already has all the settings to produce dump with different number of passes used and it's supposed to be used by developers.
`EXPLAIN SYNTAX` is supposed to be used by ordinary users, who don't know anything about query tree passes. I'd remove all these settings and reuse commits from #63402.

vdimir replied:
By default, it shows the query only after parsing, without semantic analysis. 
Actually, I started this change because sometimes it’s more convenient to see how the analyzer transforms the query tree into its AST form. 
As you mentioned earlier, expanded matches can be expected in the output of the `EXPLAIN SYNTAX` command. 
So the difference between `EXPLAIN SYNTAX` and `EXPLAIN QUERY TREE` is subtle.
My goal is to make `EXPLAIN SYNTAX` more shallow, resulting in fewer transformations while still allowing it to perform analysis if explicitly enabled. On the other hand, `EXPLAIN QUERY TREE` is intended to perform deeper analysis, though it can also be configured to perform fewer transformations if needed.

In documentation it's not overcomplicated, as soon as defaults are set https://github.com/ClickHouse/ClickHouse/pull/74536/commits/f1030a73ede4109283c44488c2bee032fb4c36f5

novikd commented on src/Interpreters/InterpreterExplainQuery.cpp:
```diff
@@ -423,69 +489,45 @@ QueryPipeline InterpreterExplainQuery::executeImpl()
+                if (explain_ok)
+                    break;
+                auto query_context_mutable = Context::createCopy(query_context);
+                query_context_mutable->setSetting("allow_experimental_analyzer", false);
```
Looks strange

vdimir replied:
Explain query tree that we reuse for explain syntax do not support queries other than `SELECT`. So for other types of queries current behavior with using old analyzer is kept.

novikd commented on src/Interpreters/InterpreterExplainQuery.cpp:
```diff
@@ -423,69 +489,45 @@ QueryPipeline InterpreterExplainQuery::executeImpl()
+                query_context = std::move(query_context_mutable);
+            }
+
+            ExplainAnalyzedSyntaxVisitor::Data data(query_context);
```
Old analyzer is used after the new one

vdimir replied:
In case of non-`SELECT` queries with are not yet supported

vdimir committed:
update doc

KochetovNicolai commented:
Let's close https://github.com/ClickHouse/ClickHouse/issues/70826 after using the new analyzer for EXPLAIN SYNTAX

vdimir commented:
The CI looks good. I’d suggest merging it as is; we can change the defaults later if needed. Re-canonizing the tests will be required again in that case, which will take some effort as mentioned above

vdimir committed:
update 03444_explain_asterisk.sql

vdimir committed:
up reference

vdimir committed:
bump sync
